pydantic/_internal/_core_metadata.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import annotations as _annotations
+
+import typing
+import warnings
+from typing import Any
+
+import typing_extensions
+from pydantic_core import CoreSchema
+from pydantic_core.core_schema import TypedDictField
+
+from ..json_schema_misc import JsonSchemaMisc
+from ._typing_extra import EllipsisType
+
+_UPDATE_CORE_SCHEMA_FIELD = 'pydantic_update_core_schema'
+_JSON_SCHEMA_MISC_FIELD = 'pydantic_json_schema_misc'
+
+_CoreMetadata = typing.Dict[str, Any]
+
+
+class UpdateCoreSchemaCallable(typing_extensions.Protocol):
+    def __call__(self, schema: CoreSchema, **kwargs: Any) -> None:
+        ...
+
+
+class CoreMetadataHandler:
+    """
+    Because the metadata field in pydantic_core is of type `Any`, we can't assume much about its contents.
+
+    This class is used to interact with the metadata field on a CoreSchema object in a consistent
+    way throughout pydantic.
+    """
+
+    def __init__(self, schema: CoreSchema | TypedDictField):
```

Comment:
@samuelcolvin I find myself making use of the `CoreSchema | TypedDictField` in a few places (mostly related to the fact that json schema annotations can find their way onto CoreSchema _or_ TypedDictField, and often the result of producing the JSON schema for a TypedDictField is something you need to retain a JSON schema $ref to). The lack of a 'type' field on TypedDictField makes it a bit of a hassle.

It's not something we _need_ to change, but it would simplify some things if it had a unique `type` field, or even better were included in the CoreSchema union. (Maybe that wouldn't work for other reasons, but I feel like I should bring it up.)